File: server/video.py
"""Generate MP4 video from scan failure delta images using ffmpeg.

Each frame shows a progress bar with running stick figure at top,
delta image fit-centered, and filename overlay at bottom.
No external Python dependencies — uses struct/zlib for PNG generation.
"""

import math
import os
import shutil
import struct
import subprocess
import tempfile
import zlib

FRAME_WIDTH = 1280
FRAME_HEIGHT = 720
FRAME_MS = 160
BG_COLOR = (26, 26, 26)
BAR_HEIGHT = 30
BAR_Y = 8


def has_ffmpeg():
    return shutil.which("ffmpeg") is not None


def generate_video(failures, output_path):
    """Generate MP4 from failure delta images."""
    if not has_ffmpeg():
        raise RuntimeError("ffmpeg not found")

    deltas = [(f, f.get("delta_path")) for f in failures if f.get("delta_path")]
    deltas = [(f, p) for f, p in deltas if os.path.isfile(p)]
    if not deltas:
        raise RuntimeError("No delta images found")

    with tempfile.TemporaryDirectory() as tmpdir:
        total_frames = len(deltas)
        trip_frames = _compute_trips(total_frames)

        overlay_paths = []
        for i in range(len(deltas)):
            progress = (i + 1) / len(deltas)
            overlay = _render_overlay(progress, i, total_frames, trip_frames)
            opath = os.path.join(tmpdir, f"overlay_{i:05d}.png")
            with open(opath, "wb") as fp:
                fp.write(overlay)
            overlay_paths.append(opath)

        # Generate frames in parallel batches using ffmpeg
        frame_paths = []
        batch_size = os.cpu_count() or 4
        cmds = []
        for i, (f, delta_path) in enumerate(deltas):
            out = os.path.join(tmpdir, f"frame_{i:05d}.png")
            frame_paths.append(out)
            filename = f.get("filename", "")
            module = f.get("module", "")
            profile = f.get("profile", "")
            diff_pct = f.get("diff_pct")
            counter = f"{i + 1} / {len(deltas)}"
            info = f"{module}{' / ' + profile if profile else ''}  -  {counter}"
            if diff_pct is not None:
                info += f"  ({diff_pct:.3f}%)"

            filename_esc = _escape_drawtext(filename)
            info_esc = _escape_drawtext(info)
            bg_hex = f"0x{BG_COLOR[0]:02x}{BG_COLOR[1]:02x}{BG_COLOR[2]:02x}"
            img_h = FRAME_HEIGHT - BAR_HEIGHT - 90

            cmds.append(
                [
                    "ffmpeg",
                    "-y",
                    "-i",
                    delta_path,
                    "-i",
                    overlay_paths[i],
                    "-filter_complex",
                    (
                        f"[0]scale={FRAME_WIDTH - 40}:{img_h}"
                        f":force_original_aspect_ratio=decrease,"
                        f"pad={FRAME_WIDTH}:{FRAME_HEIGHT}"
                        f":(ow-iw)/2:{BAR_HEIGHT + 50}+({img_h}-ih)/2"
                        f":color={bg_hex}[bg];"
                        f"[bg][1]overlay=0:0,"
                        f"drawtext=text='{filename_esc}'"
                        f":fontsize=18:fontcolor=white"
                        f":x=(w-text_w)/2:y=h-55,"
                        f"drawtext=text='{info_esc}'"
                        f":fontsize=14:fontcolor=gray"
                        f":x=(w-text_w)/2:y=h-28"
                    ),
                    "-frames:v",
                    "1",
                    out,
                ]
            )

        # Run in parallel batches
        for batch_start in range(0, len(cmds), batch_size):
            batch = cmds[batch_start : batch_start + batch_size]
            procs = [
                subprocess.Popen(
                    cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
                )
                for cmd in batch
            ]
            for p in procs:
                p.wait()

        if not frame_paths:
            raise RuntimeError("No frames generated")

        concat_path = os.path.join(tmpdir, "concat.txt")
        with open(concat_path, "w") as cf:
            for fp in frame_paths:
                cf.write(f"file '{fp}'\n")
                cf.write(f"duration {FRAME_MS / 1000:.3f}\n")
            cf.write(f"file '{frame_paths[-1]}'\n")

        subprocess.run(
            [
                "ffmpeg",
                "-y",
                "-f",
                "concat",
                "-safe",
                "0",
                "-i",
                concat_path,
                "-c:v",
                "libx264",
                "-pix_fmt",
                "yuv420p",
                "-preset",
                "fast",
                "-movflags",
                "+faststart",
                str(output_path),
            ],
            check=True,
            capture_output=True,
        )


def _render_overlay(progress, frame_idx, total_frames, trip_frames):
    """Render a transparent PNG overlay with progress bar + running stick figure."""
    bar_top = 38  # bar sits below the stick figure
    w, h = FRAME_WIDTH, bar_top + 8
    pixels = [[(0, 0, 0, 0) for _ in range(w)] for _ in range(h)]

    # Progress bar (the "ground" the figure runs on)
    for y in range(bar_top, bar_top + 4):
        for x in range(40, w - 40):
            pixels[y][x] = (60, 60, 60, 255)

    # Progress bar fill (green)
    fill_end = 40 + int((w - 80) * progress)
    for y in range(bar_top, bar_top + 4):
        for x in range(40, fill_end):
            pixels[y][x] = (34, 197, 94, 255)

    # Stick figure — only for videos with 30+ frames
    if total_frames > 30:
        fig_x = 40 + int((w - 80) * progress)
        fig_y = bar_top - 25
        _draw_stick_figure(pixels, fig_x, fig_y, w, h, frame_idx, trip_frames)

    return _encode_rgba_png(w, h, pixels)


def _compute_trips(total_frames):
    """Pick 2 random trip points in the first 50% of the video. Returns a set of frame indices."""
    import random

    if total_frames < 30:
        return set()
    half = total_frames // 2
    candidates = list(range(5, max(6, half - 5)))
    random.shuffle(candidates)
    trips = []
    for c in candidates:
        if all(abs(c - t) >= 6 for t in trips):
            trips.append(c)
        if len(trips) == 2:
            break
    result = set()
    for t in trips:
        result.update([t, t + 1, t + 2])
    return result


def _draw_stick_figure(pixels, cx, cy, w, h, frame, trip_frames):
    """Draw a running stick figure. Trips and falls occasionally."""
    phase = frame % 4
    tripping = frame in trip_frames
    color = (255, 255, 255, 255)

    def px(x, y):
        if 0 <= x < w and 0 <= y < h:
            pixels[y][x] = color

    def line(x0, y0, x1, y1):
        steps = max(abs(x1 - x0), abs(y1 - y0), 1)
        for t in range(steps + 1):
            x = x0 + (x1 - x0) * t // steps
            y = y0 + (y1 - y0) * t // steps
            px(x, y)
            px(x + 1, y)

    if tripping:
        # Determine trip phase: find which trip start this belongs to
        tp = 0
        for start in sorted(t for t in trip_frames if t - 1 not in trip_frames):
            if frame == start:
                tp = 0
            elif frame == start + 1:
                tp = 1
            elif frame == start + 2:
                tp = 2

        if tp == 0:
            # Stumbling — leaning forward
            for a in range(360):
                rad = math.radians(a)
                px(cx + 3 + int(4 * math.cos(rad)), cy + int(4 * math.sin(rad)))
            line(cx + 2, cy + 5, cx - 2, cy + 16)
            line(cx + 2, cy + 8, cx + 8, cy + 14)
            line(cx + 2, cy + 8, cx - 4, cy + 6)
            line(cx - 2, cy + 16, cx - 6, cy + 25)
            line(cx - 2, cy + 16, cx + 3, cy + 25)
        elif tp == 1:
            # Falling — horizontal
            for a in range(360):
                rad = math.radians(a)
                px(cx + 8 + int(4 * math.cos(rad)), cy + 18 + int(4 * math.sin(rad)))
            line(cx + 4, cy + 20, cx - 8, cy + 22)
            line(cx, cy + 21, cx - 4, cy + 16)
            line(cx, cy + 21, cx + 6, cy + 15)
            line(cx - 8, cy + 22, cx - 12, cy + 25)
            line(cx - 8, cy + 22, cx - 4, cy + 25)
        else:
            # Splat — flat on ground
            for a in range(360):
                rad = math.radians(a)
                px(cx + 10 + int(3 * math.cos(rad)), cy + 22 + int(3 * math.sin(rad)))
            line(cx + 7, cy + 24, cx - 8, cy + 25)
            line(cx - 2, cy + 25, cx - 6, cy + 20)
            line(cx - 2, cy + 25, cx + 4, cy + 19)
            line(cx - 8, cy + 25, cx - 10, cy + 22)
            line(cx - 8, cy + 25, cx - 5, cy + 22)
    else:
        # Normal running
        for a in range(360):
            rad = math.radians(a)
            px(cx + int(4 * math.cos(rad)), cy + int(4 * math.sin(rad)))
        line(cx, cy + 5, cx, cy + 16)
        arm_swing = [-4, -2, 2, 4][phase]
        line(cx, cy + 8, cx - 6, cy + 12 + arm_swing)
        line(cx, cy + 8, cx + 6, cy + 12 - arm_swing)
        leg_swing = [-5, -2, 3, 5][phase]
        line(cx, cy + 16, cx - 4 + leg_swing, cy + 25)
        line(cx, cy + 16, cx + 4 - leg_swing, cy + 25)


def _encode_rgba_png(width, height, pixels):
    """Encode RGBA pixel data as PNG."""
    raw = b""
    for y in range(height):
        raw += b"\x00"
        for x in range(width):
            r, g, b, a = pixels[y][x]
            raw += bytes([r, g, b, a])

    def chunk(chunk_type, data):
        c = chunk_type + data
        return (
            struct.pack(">I", len(data))
            + c
            + struct.pack(">I", zlib.crc32(c) & 0xFFFFFFFF)
        )

    return (
        b"\x89PNG\r\n\x1a\n"
        + chunk(b"IHDR", struct.pack(">IIBBBBB", width, height, 8, 6, 0, 0, 0))
        + chunk(b"IDAT", zlib.compress(raw))
        + chunk(b"IEND", b"")
    )


def _escape_drawtext(text):
    """Escape text for ffmpeg drawtext filter."""
    for ch in ("\\", "'", ":", "[", "]", ";", ","):
        text = text.replace(ch, "\\" + ch)
    return text

File: src/video.js
/**
 * Generate MP4 video from scan failure delta images using ffmpeg.
 * Progress bar + text overlay rendered as raw PNG — no drawtext needed.
 * Fully async — does not block the event loop.
 */

const fs = require('fs');
const path = require('path');
const os = require('os');
const zlib = require('zlib');
const { spawn } = require('child_process');

const FRAME_WIDTH = 1280;
const FRAME_HEIGHT = 720;
const FRAME_MS = 160;
const BG = [26, 26, 26];
const BG_HEX = `0x${BG.map(c => c.toString(16).padStart(2, '0')).join('')}`;

// Overlay: progress bar + filename + info text rendered at the bottom
const OVERLAY_H = 52; // total overlay height
const BAR_Y = 44;     // progress bar Y within overlay
const BAR_H = 4;

let _ffmpegPath = undefined;

const FFMPEG_SEARCH_PATHS = [
  '/opt/homebrew/bin/ffmpeg',
  '/usr/local/bin/ffmpeg',
  '/usr/bin/ffmpeg',
];

function findFfmpeg() {
  if (_ffmpegPath !== undefined) return _ffmpegPath;
  try {
    const { spawnSync } = require('child_process');
    const result = spawnSync('which', ['ffmpeg'], { stdio: 'pipe' });
    if (result.status === 0) {
      _ffmpegPath = result.stdout.toString().trim();
      return _ffmpegPath;
    }
  } catch {}
  for (const p of FFMPEG_SEARCH_PATHS) {
    if (fs.existsSync(p)) { _ffmpegPath = p; return p; }
  }
  _ffmpegPath = null;
  return null;
}

function hasFfmpeg() {
  return findFfmpeg() !== null;
}

function runCmd(cmd, args) {
  return new Promise((resolve, reject) => {
    const proc = spawn(cmd, args, { stdio: ['ignore', 'pipe', 'pipe'] });
    let stderr = '';
    proc.stderr.on('data', d => { stderr += d; });
    proc.on('close', code => {
      if (code === 0) resolve();
      else reject(new Error(`${path.basename(cmd)} exited ${code}: ${stderr.slice(0, 300)}`));
    });
    proc.on('error', reject);
  });
}

async function generateVideo(failures, outputPath) {
  const ffmpeg = findFfmpeg();
  if (!ffmpeg) throw new Error('ffmpeg not found');

  const deltas = failures
    .filter(f => f.delta_path && fs.existsSync(f.delta_path))
    .map(f => ({ failure: f, deltaPath: f.delta_path }));

  if (!deltas.length) throw new Error('No delta images found');

  const tmpdir = fs.mkdtempSync(path.join(os.tmpdir(), 'papastud-video-'));
  try {
    const overlayPaths = [];
    const imgH = FRAME_HEIGHT - 10 - OVERLAY_H;

    for (let i = 0; i < deltas.length; i++) {
      const { failure: f } = deltas[i];
      const progress = (i + 1) / deltas.length;
      const filename = f.filename || '';
      const mod = f.module || '';
      const profile = f.profile || '';
      const diffPct = f.diff_pct;
      const counter = `${i + 1} / ${deltas.length}`;
      let info = `${mod}${profile ? ' / ' + profile : ''}  -  ${counter}`;
      if (diffPct != null) info += `  (${diffPct.toFixed(3)}%)`;

      const overlay = renderOverlay(progress, filename, info);
      const opath = path.join(tmpdir, `overlay_${String(i).padStart(5, '0')}.png`);
      fs.writeFileSync(opath, overlay);
      overlayPaths.push(opath);
    }

    // Build frame commands — no drawtext, text is baked into overlay PNG
    const framePaths = [];
    const cmds = [];
    for (let i = 0; i < deltas.length; i++) {
      const { deltaPath } = deltas[i];
      const out = path.join(tmpdir, `frame_${String(i).padStart(5, '0')}.png`);
      framePaths.push(out);

      cmds.push([
        ffmpeg, '-y', '-i', deltaPath, '-i', overlayPaths[i],
        '-filter_complex',
        `[0]scale=${FRAME_WIDTH - 40}:${imgH}:force_original_aspect_ratio=decrease,` +
        `pad=${FRAME_WIDTH}:${FRAME_HEIGHT}:(ow-iw)/2:5+(${imgH}-ih)/2:color=${BG_HEX}[bg];` +
        `[bg][1]overlay=0:${FRAME_HEIGHT}-${OVERLAY_H}`,
        '-frames:v', '1', out,
      ]);
    }

    // Run in parallel batches (async)
    const batchSize = os.cpus().length || 4;
    for (let i = 0; i < cmds.length; i += batchSize) {
      const batch = cmds.slice(i, i + batchSize);
      await Promise.all(batch.map(cmd => runCmd(cmd[0], cmd.slice(1))));
    }

    // Concatenate
    const concatPath = path.join(tmpdir, 'concat.txt');
    let concatContent = '';
    for (const fp of framePaths) {
      concatContent += `file '${fp}'\nduration ${(FRAME_MS / 1000).toFixed(3)}\n`;
    }
    concatContent += `file '${framePaths[framePaths.length - 1]}'\n`;
    fs.writeFileSync(concatPath, concatContent);

    await runCmd(ffmpeg, [
      '-y', '-f', 'concat', '-safe', '0', '-i', concatPath,
      '-c:v', 'libx264', '-pix_fmt', 'yuv420p', '-preset', 'fast',
      '-movflags', '+faststart', outputPath,
    ]);
  } finally {
    fs.rmSync(tmpdir, { recursive: true, force: true });
  }
}

// --- Bitmap font for overlay text ---
// 5x7 pixel font for ASCII 32-126
const FONT_W = 5;
const FONT_H = 7;
const FONT_SPACING = 1;

// Compact 5x7 font glyphs — each char is 7 rows of 5-bit bitmaps
const GLYPH_DATA = {
  ' ': [0,0,0,0,0,0,0],
  '!': [4,4,4,4,4,0,4],
  '.': [0,0,0,0,0,0,4],
  ',': [0,0,0,0,0,4,8],
  ':': [0,0,4,0,0,4,0],
  '-': [0,0,0,14,0,0,0],
  '_': [0,0,0,0,0,0,31],
  '/': [1,1,2,4,8,16,16],
  '(': [2,4,8,8,8,4,2],
  ')': [8,4,2,2,2,4,8],
  '%': [24,25,2,4,8,19,3],
  '0': [14,17,19,21,25,17,14],
  '1': [4,12,4,4,4,4,14],
  '2': [14,17,1,2,4,8,31],
  '3': [14,17,1,6,1,17,14],
  '4': [2,6,10,18,31,2,2],
  '5': [31,16,30,1,1,17,14],
  '6': [6,8,16,30,17,17,14],
  '7': [31,1,2,4,8,8,8],
  '8': [14,17,17,14,17,17,14],
  '9': [14,17,17,15,1,2,12],
  'a': [0,0,14,1,15,17,15],
  'b': [16,16,30,17,17,17,30],
  'c': [0,0,14,16,16,17,14],
  'd': [1,1,15,17,17,17,15],
  'e': [0,0,14,17,31,16,14],
  'f': [6,9,8,28,8,8,8],
  'g': [0,0,15,17,15,1,14],
  'h': [16,16,22,25,17,17,17],
  'i': [4,0,12,4,4,4,14],
  'j': [2,0,6,2,2,18,12],
  'k': [16,16,18,20,24,20,18],
  'l': [12,4,4,4,4,4,14],
  'm': [0,0,26,21,21,17,17],
  'n': [0,0,22,25,17,17,17],
  'o': [0,0,14,17,17,17,14],
  'p': [0,0,30,17,30,16,16],
  'q': [0,0,15,17,15,1,1],
  'r': [0,0,22,25,16,16,16],
  's': [0,0,15,16,14,1,30],
  't': [8,8,28,8,8,9,6],
  'u': [0,0,17,17,17,19,13],
  'v': [0,0,17,17,17,10,4],
  'w': [0,0,17,17,21,21,10],
  'x': [0,0,17,10,4,10,17],
  'y': [0,0,17,17,15,1,14],
  'z': [0,0,31,2,4,8,31],
  'A': [14,17,17,31,17,17,17],
  'B': [30,17,17,30,17,17,30],
  'C': [14,17,16,16,16,17,14],
  'D': [30,17,17,17,17,17,30],
  'E': [31,16,16,30,16,16,31],
  'F': [31,16,16,30,16,16,16],
  'G': [14,17,16,23,17,17,14],
  'H': [17,17,17,31,17,17,17],
  'I': [14,4,4,4,4,4,14],
  'J': [7,2,2,2,2,18,12],
  'K': [17,18,20,24,20,18,17],
  'L': [16,16,16,16,16,16,31],
  'M': [17,27,21,17,17,17,17],
  'N': [17,17,25,21,19,17,17],
  'O': [14,17,17,17,17,17,14],
  'P': [30,17,17,30,16,16,16],
  'Q': [14,17,17,17,21,18,13],
  'R': [30,17,17,30,20,18,17],
  'S': [14,17,16,14,1,17,14],
  'T': [31,4,4,4,4,4,4],
  'U': [17,17,17,17,17,17,14],
  'V': [17,17,17,17,10,10,4],
  'W': [17,17,17,21,21,21,10],
  'X': [17,17,10,4,10,17,17],
  'Y': [17,17,10,4,4,4,4],
  'Z': [31,1,2,4,8,16,31],
  '@': [14,17,23,21,23,16,14],
  '#': [10,10,31,10,31,10,10],
};

function drawText(pixels, w, h, x, y, text, r, g, b) {
  for (let ci = 0; ci < text.length; ci++) {
    const ch = text[ci];
    const glyph = GLYPH_DATA[ch];
    if (!glyph) { x += FONT_W + FONT_SPACING; continue; }
    for (let row = 0; row < FONT_H; row++) {
      const bits = glyph[row];
      for (let col = 0; col < FONT_W; col++) {
        if (bits & (1 << (FONT_W - 1 - col))) {
          const px = x + col;
          const py = y + row;
          if (px >= 0 && px < w && py >= 0 && py < h) {
            const off = (py * w + px) * 4;
            pixels[off] = r; pixels[off + 1] = g; pixels[off + 2] = b; pixels[off + 3] = 255;
          }
        }
      }
    }
    x += FONT_W + FONT_SPACING;
  }
}

function textWidth(text) {
  return text.length * (FONT_W + FONT_SPACING) - FONT_SPACING;
}

// --- Overlay rendering ---

function renderOverlay(progress, filename, info) {
  const w = FRAME_WIDTH;
  const h = OVERLAY_H;
  const buf = Buffer.alloc(w * h * 4, 0);

  // Background strip
  for (let y = 0; y < h; y++) {
    for (let x = 0; x < w; x++) {
      const off = (y * w + x) * 4;
      buf[off] = BG[0]; buf[off + 1] = BG[1]; buf[off + 2] = BG[2]; buf[off + 3] = 255;
    }
  }

  // Info text (gray, top of overlay)
  if (info) {
    const tx = Math.floor((w - textWidth(info)) / 2);
    drawText(buf, w, h, tx, 4, info, 140, 140, 140);
  }

  // Filename (white, below info)
  if (filename) {
    const tx = Math.floor((w - textWidth(filename)) / 2);
    drawText(buf, w, h, tx, 16, filename, 255, 255, 255);
  }

  // Progress bar background
  for (let y = BAR_Y; y < BAR_Y + BAR_H; y++) {
    for (let x = 40; x < w - 40; x++) {
      const off = (y * w + x) * 4;
      buf[off] = 60; buf[off + 1] = 60; buf[off + 2] = 60; buf[off + 3] = 255;
    }
  }

  // Progress bar fill
  const fillEnd = 40 + Math.floor((w - 80) * progress);
  for (let y = BAR_Y; y < BAR_Y + BAR_H; y++) {
    for (let x = 40; x < fillEnd; x++) {
      const off = (y * w + x) * 4;
      buf[off] = 34; buf[off + 1] = 197; buf[off + 2] = 94; buf[off + 3] = 255;
    }
  }

  return encodeRgbaPng(w, h, buf);
}

function encodeRgbaPng(width, height, pixelBuf) {
  const raw = Buffer.alloc(height * (1 + width * 4));
  for (let y = 0; y < height; y++) {
    raw[y * (1 + width * 4)] = 0;
    pixelBuf.copy(raw, y * (1 + width * 4) + 1, y * width * 4, (y + 1) * width * 4);
  }
  const compressed = zlib.deflateSync(raw);

  function chunk(type, data) {
    const typeBuf = Buffer.from(type);
    const combined = Buffer.concat([typeBuf, data]);
    const len = Buffer.alloc(4); len.writeUInt32BE(data.length);
    const crc = Buffer.alloc(4); crc.writeUInt32BE(crc32(combined) >>> 0);
    return Buffer.concat([len, combined, crc]);
  }

  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8; ihdr[9] = 6;

  return Buffer.concat([
    Buffer.from([0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A]),
    chunk('IHDR', ihdr),
    chunk('IDAT', compressed),
    chunk('IEND', Buffer.alloc(0)),
  ]);
}

function crc32(buf) {
  let crc = 0xFFFFFFFF;
  for (let i = 0; i < buf.length; i++) {
    crc ^= buf[i];
    for (let j = 0; j < 8; j++) crc = (crc >>> 1) ^ (crc & 1 ? 0xEDB88320 : 0);
  }
  return crc ^ 0xFFFFFFFF;
}

module.exports = { hasFfmpeg, generateVideo };
